README.md
# Public

## Description

This project is Public site of CUP (Car Sharing Universal Platform)

Require
- [Zend Framework 2.4](https://framework.zend.com/downloads/archives) 
- [Postgre SQL 9.4.5](https://www.postgresql.org/docs/9.4/index.html)
- [Doctrine 0.9](https://www.doctrine-project.org/2015/05/05/doctrine-orm-module-release-0-9-0.html)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

This file is in [Markdown](https://www.markdownguide.org/basic-syntax/) language.

## [1.0.67] - 2020-03-24
### Changed
- Update Bankart module

## [1.0.66] - 2020-03-18
### Changed
- Updated address and phone number on footer, layout and contacts

## [1.0.65] - 2020-03-18
### Changed
- Update Bankart module (for Transaction Indicator)
- Update Public Business Module (for fix some translation)

## [1.0.64] - 2020-03-17
### Changed
- Added link to PDF file on "Terms and Conditions" message from signup form

## [1.0.63] - 2020-03-10

### Changed

- For Italian server instance, warning message for subscription suspended.

## [1.0.62] - 2020-03-06

### Added

- Removed ExtraData from the BankArt customer (see ticket [871](https://sharengo.freshdesk.com/helpdesk/tickets/871))
- Fix a translation in CustomerBonusPackages

## [1.0.61] - 2020-03-05

### Added

- Add migration for cars_bonus table

### Changed

- On CarsBonus entity, the method getUnplugEnable() manage the null value

## [1.0.60] - 2020-03-04

### Added

- Add Callback and the relative end point for Bankart payments (see ticket [871](https://sharengo.freshdesk.com/a/tickets/871) and [Asynchronous status notification](https://bankart.paymentsandbox.cloud/documentation/gateway#asynchronous-status-notification))

## [1.0.59] - 2020-03-03

### Changed

- In new-signup.phtml, add a warning message that appear if a customer select Milan/Rome fleet or try to submit

## [1.0.58] - 2020-03-03

### Changed

- Fix some bug on overlay in signup-nl1.phtml, signup-si1.phtml, signup-sk1.phtml 

## [1.0.57] - 2020-03-02

### Add

- New documentation management through views and pages are stored on database (this for FAQ, legal notes, etc.) (see [ticket 650](https://sharengo.freshdesk.com/a/tickets/650)) 

## [1.0.56] - 2020-03-02

### Changed

- fix some bug on additional services view and removed old code

## [1.0.55] - 2020-02-26

### Changed

- update pricing for Slovakia server instance
- fix problem on overlay class "sng-overlay"

## [1.0.54] - 2020-02-25

### Changed

- fix send mail on wrong payments
- update html of mail 5 ('profilo sospeso')
- fix log on pay invoice

## [1.0.53] - 2020-02-20

### Changed

- Fix Security "Sharengo web portals [WA-PT 3343]"

- Description  
  The software generates an error message that includes sensitive information about its environment, users, or associated data.

- Remediation  
  Handle exceptions internally and do not display errors containing potentially sensitive information to a user.

- Solution  
  the following lines have been changed in the "module.config.php" file
    - 'display_not_found_reason' => false,
    - 'display_exceptions' => false,

## [1.0.52] - 2020-02-19

### Added

- In additional-services.phtml add a warning message for Milan fleet

### Changed

- Changed some similar labels to optimize the translation

## [1.0.51] - 2020-02-19

### Changed

- In new-signup.phtml, add a warning message that appear if a customer select Milan fleet or try to submit

## [1.0.50] - 2020-02-18

### Changed

- changed label on new-signup2 page, for Italian drivers (B/B1 dirver license mandatory)
- restore retry payments to 60 day

## [1.0.49] - 2020-02-18

### Changed

- Improvements inside AddressController (update trip's address non payable, filter some latitude or longitude invalid)


## [1.0.48] - 2020-02-14

### Changed

- Modified AddressController for reverse geocoding via Nominatim of Open Street Map

## [1.0.47] - 2020-02-12

### Changed

- the payment retention period has been increased to 30 days instead of 60 days (ticket 494)

## [1.0.46] - 2020-02-11

### Changed

- Update Slovenian translation
- Inside public-business-module, hide the card request inside the pin page for Slovakian instance (see [ticket 781](https://sharengo.freshdesk.com/a/tickets/781))

## [1.0.45] - 2020-02-11

### Changed

- Fix Slovenian translation on first page od signup (see [ticket 767](https://sharengo.freshdesk.com/a/tickets/767))
- Update core module for block email 'bcaoo.com' (see [ticket 778](https://sharengo.freshdesk.com/a/tickets/778))

## [1.0.44] - 2020-02-07

### Changed

- In user-area for Italian version, add checkbox for optional contract conditions (news letter and general condition 1)

## [1.0.43] - 2020-02-06

### Changed

- Fix html escape char inside translations

## [1.0.42] - 2020-02-05

### Changed

- Fix some problems with "mobile" view for app in UserAreaController.php (see [ticket 683](https://sharengo.freshdesk.com/a/tickets/683))
- Add translator inside AdditionalServicesController.php
- Update po files

### Deleted

- All reference to gift-packages

## [1.0.41] - 2020-02-04

### Changed

- Update PDF privacy document for Italian version
- Update new-signup.phtml, signup-nl1.phtml, signup-si1.phtml and signup-sk1.phtml

## [1.0.40] - 2020-02-03

### Changed

- In new-signup change the checkbox form (see [ticket 662](https://sharengo.freshdesk.com/a/tickets/662))
    - privacy (mandatory), in db *customers.privacy_information*
    - vexatious clauses (mandatory), in db *customers.general_condition2*
    - new letter (optional), in db *customers.newsletter*
    - insurance info (optional), in db *customers.general_condition1*

## [1.0.39] - 2020-01-30

### Changed

- Modified rates page (see [ticket 681](https://sharengo.freshdesk.com/a/tickets/681))

## [1.0.38] - 2020-01-29

### Changed

- Modified amount of change card to 0,01 euro

## [1.0.37] - 2020-01-21

### Added 


public/js/phone.js
var countryCodes = [
    {"code": "IT", "dial_code": "+39", "name": "Italia"},
    {"code": "AT", "dial_code": "+43", "name": "Austria"},
    {"code": "BE", "dial_code": "+32", "name": "Belgium"},
    {"code": "BG", "dial_code": "+359", "name": "Bulgaria"},
    {"code": "HR", "dial_code": "+385", "name": "Croatia"},
    {"code": "CY", "dial_code": "+90", "name": "Cyprus-North"},
    {"code": "CY", "dial_code": "+357", "name": "Cyprus-South"},
    {"code": "CZ", "dial_code": "+420", "name": "CzechRepublic"},
    {"code": "DK", "dial_code": "+45", "name": "Denmark"},
    {"code": "EE", "dial_code": "+372", "name": "Estonia"},
    {"code": "FI", "dial_code": "+358", "name": "Finland"},
    {"code": "FR", "dial_code": "+33", "name": "France"},
    {"code": "DE", "dial_code": "+49", "name": "Germany"},
    {"code": "GR", "dial_code": "+30", "name": "Greece"},
    {"code": "HU", "dial_code": "+36", "name": "Hungary"},
    {"code": "IE", "dial_code": "+353", "name": "Ireland"},
    {"code": "LV", "dial_code": "+371", "name": "Latvia"},
    {"code": "LI", "dial_code": "+417", "name": "Liechtenstein"},
    {"code": "LT", "dial_code": "+370", "name": "Lithuania"},
    {"code": "LU", "dial_code": "+352", "name": "Luxembourg"},
    {"code": "MT", "dial_code": "+356", "name": "Malta"},
    {"code": "MC", "dial_code": "+377", "name": "Monaco"},
    {"code": "NL", "dial_code": "+31", "name": "Netherlands"},
    {"code": "PL", "dial_code": "+48", "name": "Poland"},
    {"code": "PT", "dial_code": "+351", "name": "Portugal"},
    {"code": "RO", "dial_code": "+40", "name": "Romania"},
    {"code": "RU", "dial_code": "+7", "name": "Russia"},
    {"code": "SM", "dial_code": "+378", "name": "SanMarino"},
    {"code": "SK", "dial_code": "+421", "name": "SlovakRepublic"},
    {"code": "SI", "dial_code": "+386", "name": "Slovenia"},
    {"code": "ES", "dial_code": "+34", "name": "Spain"},
    {"code": "SE", "dial_code": "+46", "name": "Sweden"},
    {"code": "CH", "dial_code": "+41", "name": "Switzerland"},
    {"code": "GB", "dial_code": "+44", "name": "UK"},
    {"code": "VA", "dial_code": "+379", "name": "VaticanCity"},
    {"code": "DZ", "dial_code": "+213", "name": "Algeria"},
    {"code": "AD", "dial_code": "+376", "name": "Andorra"},
    {"code": "AO", "dial_code": "+244", "name": "Angola"},
    {"code": "AI", "dial_code": "+1264", "name": "Anguilla"},
    {"code": "AG", "dial_code": "+1268", "name": "Antigua&amp;Barbuda"},
    {"code": "AR", "dial_code": "+54", "name": "Argentina"},
    {"code": "AM", "dial_code": "+374", "name": "Armenia"},
    {"code": "AW", "dial_code": "+297", "name": "Aruba"},
    {"code": "AU", "dial_code": "+61", "name": "Australia"},
    {"code": "AZ", "dial_code": "+994", "name": "Azerbaijan"},
    {"code": "BS", "dial_code": "+1242", "name": "Bahamas"},
    {"code": "BH", "dial_code": "+973", "name": "Bahrain"},
    {"code": "BD", "dial_code": "+880", "name": "Bangladesh"},
    {"code": "BB", "dial_code": "+1246", "name": "Barbados"},
    {"code": "BY", "dial_code": "+375", "name": "Belarus"},
    {"code": "BZ", "dial_code": "+501", "name": "Belize"},
    {"code": "BJ", "dial_code": "+229", "name": "Benin"},
    {"code": "BM", "dial_code": "+1441", "name": "Bermuda"},
    {"code": "BT", "dial_code": "+975", "name": "Bhutan"},
    {"code": "BO", "dial_code": "+591", "name": "Bolivia"},
    {"code": "BA", "dial_code": "+387", "name": "BosniaHerzegovina"},
    {"code": "BW", "dial_code": "+267", "name": "Botswana"},
    {"code": "BR", "dial_code": "+55", "name": "Brazil"},
    {"code": "BN", "dial_code": "+673", "name": "Brunei"},
    {"code": "BF", "dial_code": "+226", "name": "BurkinaFaso"},
    {"code": "BI", "dial_code": "+257", "name": "Burundi"},
    {"code": "KH", "dial_code": "+855", "name": "Cambodia"},
    {"code": "CM", "dial_code": "+237", "name": "Cameroon"},
    {"code": "CA", "dial_code": "+1", "name": "Canada"},
    {"code": "CV", "dial_code": "+238", "name": "CapeVerdeIslands"},
    {"code": "KY", "dial_code": "+1345", "name": "CaymanIslands"},
    {"code": "CF", "dial_code": "+236", "name": "CentralAfricanRepublic"},
    {"code": "CL", "dial_code": "+56", "name": "Chile"},
    {"code": "CN", "dial_code": "+86", "name": "China"},
    {"code": "CO", "dial_code": "+57", "name": "Colombia"},
    {"code": "KM", "dial_code": "+269", "name": "Comoros"},
    {"code": "CG", "dial_code": "+242", "name": "Congo"},
    {"code": "CK", "dial_code": "+682", "name": "CookIslands"},
    {"code": "CR", "dial_code": "+506", "name": "CostaRica"},
    {"code": "CU", "dial_code": "+53", "name": "Cuba"},
    {"code": "DJ", "dial_code": "+253", "name": "Djibouti"},
    {"code": "DM", "dial_code": "+1809", "name": "Dominica"},
    {"code": "DO", "dial_code": "+1809", "name": "DominicanRepublic"},
    {"code": "EC", "dial_code": "+593", "name": "Ecuador"},
    {"code": "EG", "dial_code": "+20", "name": "Egypt"},
    {"code": "SV", "dial_code": "+503", "name": "ElSalvador"},
    {"code": "GQ", "dial_code": "+240", "name": "EquatorialGuinea"},
    {"code": "ER", "dial_code": "+291", "name": "Eritrea"},
    {"code": "ET", "dial_code": "+251", "name": "Ethiopia"},
    {"code": "FK", "dial_code": "+500", "name": "FalklandIslands"},
    {"code": "FO", "dial_code": "+298", "name": "FaroeIslands"},
    {"code": "FJ", "dial_code": "+679", "name": "Fiji"},
    {"code": "GF", "dial_code": "+594", "name": "FrenchGuiana"},
    {"code": "PF", "dial_code": "+689", "name": "FrenchPolynesia"},
    {"code": "GA", "dial_code": "+241", "name": "Gabon"},
    {"code": "GM", "dial_code": "+220", "name": "Gambia"},
    {"code": "GE", "dial_code": "+7880", "name": "Georgia"},
    {"code": "GH", "dial_code": "+233", "name": "Ghana"},
    {"code": "GI", "dial_code": "+350", "name": "Gibraltar"},
    {"code": "GL", "dial_code": "+299", "name": "Greenland"},
    {"code": "GD", "dial_code": "+1473", "name": "Grenada"},
    {"code": "GP", "dial_code": "+590", "name": "Guadeloupe"},
    {"code": "GU", "dial_code": "+671", "name": "Guam"},
    {"code": "GT", "dial_code": "+502", "name": "Guatemala"},
    {"code": "GN", "dial_code": "+224", "name": "Guinea"},
    {"code": "GW", "dial_code": "+245", "name": "Guinea-Bissau"},
    {"code": "GY", "dial_code": "+592", "name": "Guyana"},
    {"code": "HT", "dial_code": "+509", "name": "Haiti"},
    {"code": "HN", "dial_code": "+504", "name": "Honduras"},
    {"code": "HK", "dial_code": "+852", "name": "HongKong"},
    {"code": "IS", "dial_code": "+354", "name": "Iceland"},
    {"code": "IN", "dial_code": "+91", "name": "India"},
    {"code": "ID", "dial_code": "+62", "name": "Indonesia"},
    {"code": "IQ", "dial_code": "+964", "name": "Iraq"},
    {"code": "IR", "dial_code": "+98", "name": "Iran"},
    {"code": "IL", "dial_code": "+972", "name": "Israel"},
    {"code": "JM", "dial_code": "+1876", "name": "Jamaica"},
    {"code": "JP", "dial_code": "+81", "name": "Japan"},
    {"code": "JO", "dial_code": "+962", "name": "Jordan"},
    {"code": "KZ", "dial_code": "+7", "name": "Kazakhstan"},
    {"code": "KE", "dial_code": "+254", "name": "Kenya"},
    {"code": "KI", "dial_code": "+686", "name": "Kiribati"},
    {"code": "KP", "dial_code": "+850", "name": "Korea-North"},
    {"code": "KR", "dial_code": "+82", "name": "Korea-South"},
    {"code": "KW", "dial_code": "+965", "name": "Kuwait"},
    {"code": "KG", "dial_code": "+996", "name": "Kyrgyzstan"},
    {"code": "LA", "dial_code": "+856", "name": "Laos"},
    {"code": "LB", "dial_code": "+961", "name": "Lebanon"},
    {"code": "LS", "dial_code": "+266", "name": "Lesotho"},
    {"code": "LR", "dial_code": "+231", "name": "Liberia"},
    {"code": "LY", "dial_code": "+218", "name": "Libya"},
    {"code": "MO", "dial_code": "+853", "name": "Macao"},
    {"code": "MK", "dial_code": "+389", "name": "Macedonia"},
    {"code": "MG", "dial_code": "+261", "name": "Madagascar"},
    {"code": "MW", "dial_code": "+265", "name": "Malawi"},
    {"code": "MY", "dial_code": "+60", "name": "Malaysia"},
    {"code": "MV", "dial_code": "+960", "name": "Maldives"},
    {"code": "ML", "dial_code": "+223", "name": "Mali"},
    {"code": "MH", "dial_code": "+692", "name": "MarshallIslands"},
    {"code": "MQ", "dial_code": "+596", "name": "Martinique"},
    {"code": "MR", "dial_code": "+222", "name": "Mauritania"},
    {"code": "YT", "dial_code": "+269", "name": "Mayotte"},
    {"code": "MX", "dial_code": "+52", "name": "Mexico"},
    {"code": "FM", "dial_code": "+691", "name": "Micronesia"},
    {"code": "MD", "dial_code": "+373", "name": "Moldova"},
    {"code": "MN", "dial_code": "+976", "name": "Mongolia"},
    {"code": "MS", "dial_code": "+1664", "name": "Montserrat"},
    {"code": "MA", "dial_code": "+212", "name": "Morocco"},
    {"code": "MZ", "dial_code": "+258", "name": "Mozambique"},
    {"code": "MN", "dial_code": "+95", "name": "Myanmar"},
    {"code": "NA", "dial_code": "+264", "name": "Namibia"},
    {"code": "NR", "dial_code": "+674", "name": "Nauru"},
    {"code": "NP", "dial_code": "+977", "name": "Nepal"},
    {"code": "NC", "dial_code": "+687", "name": "NewCaledonia"},
    {"code": "NZ", "dial_code": "+64", "name": "NewZealand"},
    {"code": "NI", "dial_code": "+505", "name": "Nicaragua"},
    {"code": "NE", "dial_code": "+227", "name": "Niger"},
    {"code": "NG", "dial_code": "+234", "name": "Nigeria"},
    {"code": "NU", "dial_code": "+683", "name": "Niue"},
    {"code": "NF", "dial_code": "+672", "name": "NorfolkIslands"},
    {"code": "NP", "dial_code": "+670", "name": "NorthernMarianas"},
    {"code": "NO", "dial_code": "+47", "name": "Norway"},
    {"code": "OM", "dial_code": "+968", "name": "Oman"},
    {"code": "PK", "dial_code": "+92", "name": "Pakistan"},
    {"code": "PW", "dial_code": "+680", "name": "Palau"},
    {"code": "PA", "dial_code": "+507", "name": "Panama"},
    {"code": "PG", "dial_code": "+675", "name": "PapuaNewGuinea"},
    {"code": "PY", "dial_code": "+595", "name": "Paraguay"},
    {"code": "PE", "dial_code": "+51", "name": "Peru"},
    {"code": "PH", "dial_code": "+63", "name": "Philippines"},
    {"code": "PR", "dial_code": "+1787", "name": "PuertoRico"},
    {"code": "QA", "dial_code": "+974", "name": "Qatar"},
    {"code": "RE", "dial_code": "+262", "name": "Reunion"},
    {"code": "RW", "dial_code": "+250", "name": "Rwanda"},
    {"code": "ST", "dial_code": "+239", "name": "SaoTome&amp;Principe"},
    {"code": "SA", "dial_code": "+966", "name": "SaudiArabia"},
    {"code": "SN", "dial_code": "+221", "name": "Senegal"},
    {"code": "CS", "dial_code": "+381", "name": "Serbia"},
    {"code": "SC", "dial_code": "+248", "name": "Seychelles"},
    {"code": "SL", "dial_code": "+232", "name": "SierraLeone"},
    {"code": "SG", "dial_code": "+65", "name": "Singapore"},
    {"code": "SB", "dial_code": "+677", "name": "SolomonIslands"},
    {"code": "SO", "dial_code": "+252", "name": "Somalia"},
    {"code": "ZA", "dial_code": "+27", "name": "SouthAfrica"},
    {"code": "LK", "dial_code": "+94", "name": "SriLanka"},
    {"code": "SH", "dial_code": "+290", "name": "St.Helena"},
    {"code": "KN", "dial_code": "+1869", "name": "St.Kitts"},
    {"code": "SC", "dial_code": "+1758", "name": "St.Lucia"},
    {"code": "SR", "dial_code": "+597", "name": "Suriname"},
    {"code": "SD", "dial_code": "+249", "name": "Sudan"},
    {"code": "SZ", "dial_code": "+268", "name": "Swaziland"},
    {"code": "SY", "dial_code": "+963", "name": "Syria"},
    {"code": "TW", "dial_code": "+886", "name": "Taiwan"},
    {"code": "TJ", "dial_code": "+992", "name": "Tajikistan"},
    {"code": "TH", "dial_code": "+66", "name": "Thailand"},
    {"code": "TG", "dial_code": "+228", "name": "Togo"},
    {"code": "TO", "dial_code": "+676", "name": "Tonga"},
    {"code": "TT", "dial_code": "+1868", "name": "Trinidad&amp;Tobago"},
    {"code": "TN", "dial_code": "+216", "name": "Tunisia"},
    {"code": "TR", "dial_code": "+90", "name": "Turkey"},
    {"code": "TM", "dial_code": "+993", "name": "Turkmenistan"},
    {"code": "TC", "dial_code": "+1649", "name": "Turks&amp;CaicosIslands"},
    {"code": "TV", "dial_code": "+688", "name": "Tuvalu"},
    {"code": "UG", "dial_code": "+256", "name": "Uganda"},
    {"code": "UA", "dial_code": "+380", "name": "Ukraine"},
    {"code": "AE", "dial_code": "+971", "name": "UnitedArabEmirates"},
    {"code": "UY", "dial_code": "+598", "name": "Uruguay"},
    {"code": "US", "dial_code": "+1", "name": "USA"},
    {"code": "UZ", "dial_code": "+998", "name": "Uzbekistan"},
    {"code": "VU", "dial_code": "+678", "name": "Vanuatu"},
    {"code": "VE", "dial_code": "+58", "name": "Venezuela"},
    {"code": "VN", "dial_code": "+84", "name": "Vietnam"},
    {"code": "VG", "dial_code": "+1", "name": "VirginIslands-British"},
    {"code": "VI", "dial_code": "+1", "name": "VirginIslands-US"},
    {"code": "WF", "dial_code": "+681", "name": "Wallis&amp;Futuna"},
    {"code": "YE", "dial_code": "+969", "name": "Yemen North"},
    {"code": "YE", "dial_code": "+967", "name": "Yemen South"},
    {"code": "ZM", "dial_code": "+260", "name": "Zambia"},
    {"code": "ZW", "dial_code": "+263", "name": "Zimbabwe"}
];

var phoneJs = function (mobile) {

    var result = mobile;
    var prefix = "";
    var dialcode = "39";

    if (mobile.charAt(0) == '+') {
        var country = getCountry(mobile);
        if (country !== null) {
            prefix = country["dial_code"];
            dialcode = prefix.substring(1, prefix.length); 
            result = mobile.replace(prefix, "");
        }
    }

    return {"dial_code": dialcode, "number": result};
};

var getCountry = function (mobile) {

    var result = null;
    var country = null;

    for (i = 1; i < 5; i++) {
        var firstPart = mobile.substring(0, i + 1);
        var count = 0;
        for (var key in countryCodes) {
            country = countryCodes[key];
            if (country["dial_code"] !== null) {
                dialCode = country["dial_code"];
                if (dialCode === firstPart) {
                    count++;
                    result = country;
                }
            }
        }
        if (count == 1)
            break;
        else{
            if (mobile.substring(0, 2) === "+1")
                return result = {"code": "US", "dial_code": "+1", "name": "USA"};
            else
                result = null;
        }
    }

    return result;
};
